### PR TITLE
Infer a good name for patterns that represent case-insensitive keywords

### LINKED ISSUE #51: Infer the name of case-insensitive keywords e.g. /[Ii][Ff]/ -> if

### PR DESCRIPTION
**Pull request by mjambon**
Closes #51 

This changes type names for some patterns in the generated `CST` module. I expect this will benefit Dockerfile, Apex, and any language that uses case-insensitive keywords.

### Security

- [x] Change has no security implications (otherwise, ping the security team)

### CODE REVIEW

**Comment by mjambon** [comment]
I ran the following commands in ocaml-tree-sitter-semgrep to find out which languages are affected.

Step 1: Check that we have grammars in JSON form for all the languages, because some patterns are generated programmatically from JavaScript.

```
$ for x in semgrep-*; do wc -l $x/src/grammar.json || wc -l $x/*/src/grammar.json ; done
```

Step 2: Look for regexps matching things like `"[Aa][Bb]"`

```
~/ocaml-tree-sitter-semgrep/lang/semgrep-grammars/src $ find semgrep-* -name '*.json' | xargs grep -P '"(\[..\])+"' | sed -e 's/ \+/ /g' | sort -u | cut -d: -f1 | uniq -c | sort -nr
    199 semgrep-sfapex/apex/src/grammar.json
    151 semgrep-sfapex/sosl/src/grammar.json
    149 semgrep-sqlite/src/grammar.json
    136 semgrep-sfapex/soql/src/grammar.json
     54 semgrep-php/src/grammar.json
     19 semgrep-dockerfile/src/grammar.json
      7 semgrep-kotlin/src/grammar.json
      5 semgrep-swift/src/grammar.json
      2 semgrep-python/src/grammar.json
      2 semgrep-ocaml/src/grammar.json
      2 semgrep-go/src/grammar.json
      1 semgrep-ruby/src/grammar.json
      1 semgrep-java/src/grammar.json
      1 semgrep-html/src/grammar.json
      1 semgrep-elixir/src/grammar.json
      1 semgrep-dart/src/grammar.json
```
(truncate the pipeline after `sort -u` to see the matches)

The most affected languages are:
* OK: sfapex is what I'm starting to work on.
* OK: sqlite hasn't been integrated in semgrep yet.
* php
* dockerfile

The changes for the other languages are rather small. I'll let their respective authors deal with it when they update the grammar. For example, the affected patterns for Kotlin are:
```
semgrep-kotlin/src/grammar.json: "value": "[+-]"
semgrep-kotlin/src/grammar.json: "value": "[01]"
semgrep-kotlin/src/grammar.json: "value": "[bB]"
semgrep-kotlin/src/grammar.json: "value": "[eE]"
semgrep-kotlin/src/grammar.json: "value": "[fF]"
semgrep-kotlin/src/grammar.json: "value": "[uU]"
semgrep-kotlin/src/grammar.json: "value": "[xX]"
```

In conclusion, I'll make the necessary changes for PHP and for Dockerfile because it's easier done now than later.

**Comment by mjambon** [comment]
There wasn't much to do to adjust to these changes in the semgrep repo after all. See the 2 PRs shown above.

File: src/gen/lib/Pattern_name.mll [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        Hashtbl.add tbl x ();
+        true
+      )
+    ) xs
+
+  let normalize_char = function
+  | '-' -> '_'
+  | c -> Char.lowercase_ascii c
+
+  let lowercase_singleton_of_cset s =
```
Comment (nmote) [inline src/gen/lib/Pattern_name.mll:23]: For consistency with `normalize_char`, and since we also modify `-`, maybe `normalized_singleton_of_cset`?

File: src/gen/lib/Pattern_name.mll [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  let infer s =
+    let lexbuf = Lexing.from_string s in
+    let elements = sequence lexbuf in
+    let chars =
+      try Some (List.map (function Char c -> c | Other -> raise Exit) elements)
+      with Exit -> None
+    in
+    match chars with
+    | Some chars ->
+        let chars = List.map (function '-' -> '_' | c -> c) chars in
```
Comment (nmote) [inline src/gen/lib/Pattern_name.mll:62]: This is redundant, right? This transformation should have already been done in `normalize_char` above?

**Review by nmote — APPROVED** [review]
Nice! I like the use of the small generated parser for this. Very concise.

File: src/gen/lib/Pattern_name.mll (hunk 2 above)
Comment (mjambon) [reply src/gen/lib/Pattern_name.mll:62]: oops, yes. It needs cleanup.

File: src/gen/lib/Pattern_name.mll (hunk 1 above)
Comment (mjambon) [reply src/gen/lib/Pattern_name.mll:23]: thanks!
